File: library/src/main/java/me/proxer/library/api/messenger/ConferenceModificationEndpoint.java
package me.proxer.library.api.messenger;

import lombok.experimental.Accessors;
import me.proxer.library.api.Endpoint;
import me.proxer.library.api.ProxerCall;

/**
 * Endpoint for modifying a conference in various ways.
 * <p>
 * Possible modifications are:
 * - Mark as read
 * - Mark as unread
 * - Block
 * - Unblock
 * - Mark as favourite
 * - Unmark as favourite
 *
 * @author Ruben Gees
 */
@Accessors(fluent = true)
public class ConferenceModificationEndpoint implements Endpoint<Void> {

    private final InternalApi internalApi;

    private final String id;
    private final ConferenceModification modification;

    ConferenceModificationEndpoint(final InternalApi internalApi, final String id,
                                   final ConferenceModification modification) {
        this.internalApi = internalApi;
        this.id = id;
        this.modification = modification;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public ProxerCall<Void> build() {
        switch (modification) {
            case READ:
                return internalApi.markConferenceAsRead(id);
            case UNREAD:
                return internalApi.unmarkConferenceAsRead(id);
            case BLOCK:
                return internalApi.markConferenceAsBlocked(id);
            case UNBLOCK:
                return internalApi.unmarkConferenceAsBlocked(id);
            case FAVOUR:
                return internalApi.markConferenceAsFavorite(id);
            case UNFAVOUR:
                return internalApi.unmarkConferenceAsFavorite(id);
            default:
                throw new IllegalArgumentException("Illegal modification: " + modification);
        }
    }

    enum ConferenceModification {
        READ,
        UNREAD,
        BLOCK,
        UNBLOCK,
        FAVOUR,
        UNFAVOUR
    }
}

File: library/src/main/kotlin/me/proxer/library/api/messenger/ConferenceModificationEndpoint.kt
package me.proxer.library.api.messenger

import me.proxer.library.ProxerCall
import me.proxer.library.api.Endpoint
import me.proxer.library.api.messenger.ConferenceModificationEndpoint.ConferenceModification.BLOCK
import me.proxer.library.api.messenger.ConferenceModificationEndpoint.ConferenceModification.FAVOUR
import me.proxer.library.api.messenger.ConferenceModificationEndpoint.ConferenceModification.READ
import me.proxer.library.api.messenger.ConferenceModificationEndpoint.ConferenceModification.UNBLOCK
import me.proxer.library.api.messenger.ConferenceModificationEndpoint.ConferenceModification.UNFAVOUR
import me.proxer.library.api.messenger.ConferenceModificationEndpoint.ConferenceModification.UNREAD

/**
 * Endpoint for modifying a conference in various ways.
 *
 * Possible modifications are:
 * - Mark as read
 * - Mark as unread
 * - Block
 * - Unblock
 * - Mark as favourite
 * - Unmark as favourite
 *
 * @author Ruben Gees
 */
class ConferenceModificationEndpoint internal constructor(
    private val internalApi: InternalApi,
    private val id: String,
    private val modification: ConferenceModification
) : Endpoint<Unit?> {

    override fun build(): ProxerCall<Unit?> {
        return when (modification) {
            READ -> internalApi.markConferenceAsRead(id)
            UNREAD -> internalApi.unmarkConferenceAsRead(id)
            BLOCK -> internalApi.markConferenceAsBlocked(id)
            UNBLOCK -> internalApi.unmarkConferenceAsBlocked(id)
            FAVOUR -> internalApi.markConferenceAsFavorite(id)
            UNFAVOUR -> internalApi.unmarkConferenceAsFavorite(id)
        }
    }

    internal enum class ConferenceModification {
        READ, UNREAD, BLOCK, UNBLOCK, FAVOUR, UNFAVOUR
    }
}
